Make repair/rearm of ground units queueable

reaperrr opened the pull request:
Moves `Repairable(Near)` activity handling to `Resupply` activity, plus a few fixes.

Implements first step of https://github.com/OpenRA/OpenRA/issues/15023#issuecomment-489332380

Fixes #16462.
Fixes #16463.
Fixes #16761.
Fixes #16492.

reaperrr committed:
Fix Resupply regression

The work-around was originally written with ground units in mind
and caused issues with aircraft.

pchote commented on OpenRA.Mods.Common/Traits/Repairable.cs:
```diff
@@ -123,41 +123,19 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
-
-				self.QueueActivity(new WaitForTransport(self, activities));
-				TryCallTransport(self, order.Target, new CallFunc(() => AfterReachActivities(self, order, movement)));
+				self.QueueActivity(new MoveToResupplier(self, order.Target));
```
We really want to be queuing `Resupply` here, and that checks whether it is in the correct position and moves if needed. See https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Activities/HarvestResource.cs#L62 for a recent example

reaperrr committed:
Move movement to resupplier inside Resupply activity

From Repairable(Near).

reaperrr committed:
Move AmmoPool RemainingTicks reset to Rearmable

This is a better place to do this than the Resupply activity.

reaperrr commented:
Updated.

teinarss commented:
We are missing the target lines when ordering repairing. Is this intended and will be fixed with #16549?

reaperrr committed:
Make Repairable(Near) implement interfaces explicitly

reaperrr commented:
That wasn't intended, no.

Updated.

teinarss reviewed (dismissed):
LGTM

abcdefg30 reviewed (dismissed):
Helicopters in RA don't take off at all if you queue a reload and repair activity.

reaperrr commented:
> Helicopters in RA don't take off at all if you queue a reload and repair activity.

Not this PR's fault and intentional, see discussion in #16660. We can still change that back after the first playtest if the feedback on that is negative.

As far as this PR is concerned, manually set `AbortOnResupply: false` if you need a test case for aircraft (works fine according to my testing).

reaperrr commented:
As discussed on IRC, fixing that in a non-work-around way requires #16509 to be finished and merged first.

pchote commented:
#16509 has been merged.

pchote commented:
> 23:24 <+pchote> re the comments in #16721, I have a sinking feeling that we may end up having to rework AbortOnResupply before the end of the playtest series
23:24 <@orabot> Pull request #16721 (open) by reaperrr: Make repair/rearm of ground units queueable | http://bugs.openra.net/16721
23:25 <+reaperrr> ugh but yeah, am not that surprised
23:25 <+pchote> if you have a damaged aircraft and want to queue a repair/reload and then move somewhere that isn't the rallypoint then this really should be possible
23:25 <+pchote> and this is independent from whether they break off after a single attack run
23:26 <+pchote> tovl reworks the flag a bit in #16481
23:26 <@orabot> Pull request #16481 (open) by tovl: Move ChildActivity handling into base Activity class | http://bugs.openra.net/16481
23:27 <+pchote> I expect that after that has been merged that we'll need to take it further and probably move it to AttackAircraft
23:27 <+pchote> have it abort FlyAttack only, not the whole quue
23:27 <+pchote> *queue
23:28 <+reaperrr> yeah

reaperrr committed:
Move AbortOnResupply to AttackAircraft

Additionally, if AbortOnResupply is set to 'true',
abort FlyAttack right away when queueing ReturnToBase.

reaperrr commented:
Updated. Fingers crossed that the last commit might already be enough.

pchote commented:
Needs a rebase after #16481 was merged.

reaperrr commented:
Rebased.

pchote commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -111,11 +111,17 @@ public override bool Tick(Actor self)
 			{
-				QueueChild(new ReturnToBase(self));
-				return aircraft.Info.AbortOnResupply;
+				// End this activity after reload has been queued if AbortOnResupply is true,
```
This is what the `return attackAircraft.Info.AbortOnResupply;` should already be doing. Are you sure these changes are still needed after the rebase?

reaperrr replied:
No, but I also wasn't sure what happens to the child when the parent activity ends. It seems to work as intended though, so I'll drop my change.

reaperrr commented:
Updated. Also added an update rule for mods that intentionally set `AbortOnResupply` to `false`.

pchote commented:
Needs a rebase again / still?

pchote commented on OpenRA.Mods.Common/Activities/Resupply.cs:
```diff
@@ -56,46 +61,77 @@ public Resupply(Actor self, Actor host, WDist closeEnough)
-				|| (closeEnough.LengthSquared > 0 && !host.IsInRange(self.CenterPosition, closeEnough)))
+					return true;
+				}
+				else if (self.Info.HasTraitInfo<MobileInfo>())
```
Can save some trait(info) lookups here by doing
```
var move = self.Trait<IMove>();
var aircraft = move as Aircraft;
```

above, and then using `move` instead of `MobileInfo` and the lookups below.

pchote commented on OpenRA.Mods.Common/Activities/Resupply.cs:
```diff
@@ -56,46 +61,77 @@ public Resupply(Actor self, Actor host, WDist closeEnough)
 
 				return true;
 			}
+			else if (activeResupplyTypes != 0 && !self.Info.HasTraitInfo<AircraftInfo>() &&
```
can reuse `aircraft` instead of querying `AircraftInfo`

pchote commented on OpenRA.Mods.Common/Activities/Resupply.cs:
```diff
@@ -112,6 +148,16 @@ public override bool Tick(Actor self)
+				{
+					var rp = host.Actor.TraitOrDefault<RallyPoint>();
+					if (rp != null)
+						Queue(self.Trait<IMove>().MoveTo(rp.Location, host.Actor));
```
and then reuse `move` here.

pchote commented on OpenRA.Mods.Common/Traits/Repairable.cs:
```diff
@@ -99,65 +97,44 @@ bool CanRearm()
-				if (!CanRepairAt(order.Target.Actor) || (!CanRepair() && !CanRearm()))
-					return;
+			// Aircraft handle Repair orders directly in the Aircraft trait
+			if (self.Info.HasTraitInfo<AircraftInfo>())
```
You could cache a `bool isAircraft` from `INotifyCreated.Created` above.

...or maybe remove this Aircraft special case completely, like we discussed the other day? I assume that didn't go so well...

reaperrr replied:
> I assume that didn't go so well...

Nope, another can of worms. Going by experience, this looks like something that could easily take 4-8 weeks to get right (delays in reviewing and updating calculated in).

reaperrr commented:
> Needs a rebase again / still?

Again. The FlightDynamics PR was merged after I had rebased this one.

Updated and rebased.

pchote commented:
The " Fix Resupply regression" commit references `AircraftInfo.TakeOffOnResupply` instead of `AircraftInfo.FlightDynamics`. Can you please fix that commit so it compiles cleanly for the sake of future bisects?

pchote commented on OpenRA.Mods.Common/Activities/Resupply.cs:
```diff
@@ -119,10 +153,19 @@ public override bool Tick(Actor self)
 					aircraft.AllowYieldingReservation();
 
+				var actorHasNextActivity = self.CurrentActivity.NextActivity != null;
+				if (!actorHasNextActivity && repairableNear == null)
```
TIL that `RepairableNear` was never hooked up to rally points. Can we please fix this while we are here?

reaperrr replied:
Personally, I found the move to rallypoint a bit annoying for RepairableNear, because
- the repaired actor isn't blocking the only repair slot if it doesn't move
- depending on where it was repaired, it may have to move around the entire building to reach the rallypoint
- as ship yards/sub pens are also production facilities, things can get crowded at the rp even faster if repaired actors move there as well

I don't really mind changing it anyway, just wanted to point this out.

pchote reviewed (commented):
One more request, then LGTM

reaperrr commented:
Updated.

> Can you please fix that commit so it compiles cleanly for the sake of future bisects?

Done.

fruestueck commented:
Fixes #16761 

I had a small bug where a vehicle first moved to the top-left cell obove the RA service depot but I can't reproduce it. Could be nothing.

abcdefg30 reviewed (dismissed):
Something seems to be going on with `RepairableNear`:
![danceDestroy](https://user-images.githubusercontent.com/7704140/60915433-1ef4ed80-a28c-11e9-8dff-05359c93a807.gif)
Testcase: Issue a stop order while repairing. The unit will not respond to any other orders.

Otherwise the PR looks good to me.

reaperrr commented:
Fixed.
The problem was that the `MoveToTarget` on cancellation is bogus for `RepairableNear` because a) it doesn't need to leave the host and b) it would always remain within `closeEnough` range, re-queueing the move infinitely.

Also merged the fix and the previously separate `RepairableNear` commit into 
4984788.

pchote commented:
Unfortunately `RepairableNear` is still not behaving correctly.

Testcase:
1) Damage a boat
2) order it to repair + queue a move somewhere afterwards.
3) issue a non-queued move order while it is repairing.

Expected behaviour: repair and queued move are cancelled, boat moves immediately to ordered location.
Actual behaviour: repair continues until completed, boat then moves to the original queued location.

reaperrr commented:
I have no friggin' clue why. The way I understand the updated Activity Tick code, `return true` should be equal to the old `return NextActivity`, and `Mobile` calls `self.CancelActivity()` on non-queued move orders, so it doesn't make any sense to me that the resupply continues, nor that the queued orders aren't cancelled.

pchote commented:
An important detail from `Activity.Tick` is
https://github.com/OpenRA/OpenRA/blob/1c03fb9e516e0690a2008749d754ad453e0dd733/OpenRA.Game/Activities/Activity.cs#L106-L108

The `Resupply` activity has a custom `Cancel` override that prevents `base.Cancel` (and therefore the child activities) from being cancelled if there are queued activities. This forces the child activities to be uncancellable and run until completion.

tovl commented:
The `Cancel` override needs to be removed. I don't even remember what I was thinking there. I guess this was supposed to help with the issue of `Resupply` being cancelled by a stop command, but it is complete bogus.

reaperrr committed:
Remove Resupply Cancel override

This became detrimental to actor responsiveness
while resupplying.

reaperrr committed:
Improve handling of finished/cancelled Resupply

reaperrr commented:
Updated, left the updates in separate commits (for now).

tovl commented on OpenRA.Mods.Common/Activities/Resupply.cs:
```diff
@@ -56,46 +65,86 @@ public Resupply(Actor self, Actor host, WDist closeEnough)
+				{
+					aircraft.AllowYieldingReservation();
+					if (aircraft.Info.FlightDynamics.HasFlag(FlightDynamic.TakeOffOnResupply))
+						Queue(new TakeOff(self));
```
`ReturnToBase` already takes care of queueing the `TakeOff` so this seems a bit redundant. Also, is there a reason why this is not queued as a childactivity? This wouldn't work correctly when other activities are queued after `Resupply.`

pchote replied:
I've lost track of the specifics about `ReturnToBase` and taking off, but having the `TakeOff` `Queue`d after other user-ordered activities is definitely an issue.

reaperrr commented:
Updated.

tovl commented on OpenRA.Mods.Common/Activities/Resupply.cs:
```diff
@@ -78,18 +78,29 @@ public override bool Tick(Actor self)
 
-					return true;
+					// ReturnToBase only handles TakeOffOnResupply when Resupply finishes, not when it's cancelled prematurely
+					if (aircraft.Info.FlightDynamics.HasFlag(FlightDynamic.TakeOffOnResupply))
```
The `TakeOff` should be suppressed when `Resupply` is called by `ReturnToBase` with `alwaysLand = true`. This can be done by adding an `alwaysStay` parameter to the constructor and having `ReturnToBase` pass `alwaysLand` to that.

tovl commented on OpenRA.Mods.Common/Activities/Resupply.cs:
```diff
@@ -148,16 +159,16 @@ public override bool Tick(Actor self)
 
 			if (activeResupplyTypes == 0)
 			{
+				// Resupply properly finished, so ReturnToBase will handle the TakeOffOnResupply
```
`TakeOffOnResupply` is now handled in two different places. It might be better to remove the `TakeOff` handling from `ReturnToBase` and handle it here. I moved the `TakeOff` to `ReturnToBase` to handle the `alwaysLand` case, but that can be handled by passing the boolean down to `Resupply` instead (per the above comment).

tovl commented on OpenRA.Mods.Common/Activities/Resupply.cs:
```diff
@@ -78,18 +78,29 @@ public override bool Tick(Actor self)
 				if (aircraft != null)
```
The entire block handling reservation, takeoff and rallypoints is now repeated twice, so it would make sense to pull this into a method so as to not repeat ourselves.

tovl commented on OpenRA.Mods.Common/Activities/Resupply.cs:
```diff
@@ -78,18 +78,29 @@ public override bool Tick(Actor self)
+					// If there's no next activity, move to rallypoint if available, else just leave host if Repairable.
+					// Do nothing if RepairableNear (RepairableNear actors don't enter their host and will likely remain within closeEnough).
+					// If there's a next activity and we're not RepairableNear, first leave host if the next activity is not a Move.
+					if (self.CurrentActivity.NextActivity == null)
```
It would be nice if ground units also had the option to stay on the resupplier after resupplying. This could be done by replacing `TakeOffOnResupply` with `LeaveOnRepair` and `LeaveOnRearm` which would be parameters of the `Repairable` and `Rearmable` trait respectively. This would be a step in separating the resupply logic from `Aircraft`. Having separate booleans for repair and rearm has the added benefit of allowing aircraft to takeoff after repairing on the repair pad but not after rearming on the airfield.

reaperrr replied:
I agree, but I'd like to leave that step for a follow-up, we're already quite late with the playtest.

tovl replied:
Yes, this is more something for a follow-up than something that needs to happen right now.

reaperrr committed:
Fixed comment typo in Aircraft

reaperrr commented:
Updated.

tovl commented on OpenRA.Mods.Common/Activities/Resupply.cs:
```diff
@@ -148,22 +137,39 @@ public override bool Tick(Actor self)
+			{
+				aircraft.AllowYieldingReservation();
+
+				// ReturnToBase only handles TakeOffOnResupply when Resupply finishes, not when it's cancelled prematurely
```
This comment can now be removed.

reaperrr replied:
Done.

tovl commented:
This does not seem to fix #16463.

reaperrr committed:
Make aircraft always take off after repair

Reservable logic doesn't handle repairs, and we
don't want aircraft to block repair bays etc. until it does.

reaperrr commented:
> This does not seem to fix #16463.

Right, forgot that this would require more changes.
I've added a hacky short-term fix for the upcoming release, let's leave the proper fix (`TakeOffOn` flag list?) for a follow-up.

tovl commented:
The fix works. However, aircraft with `TakeOffOnResupply == false` will not move to the rally point when taking off from the repair bay.

I also noticed that when an aircraft with `TakeOffOnResupply == true` is made to resupply using the `force-enter` command and the `stop` command is given it will forget to supress the `TakeOff`. This is due to the `Resupply` activity being reqeued without the `stayOnResupplier` parameter.

reaperrr commented:
> The fix works. However, aircraft with TakeOffOnResupply == false will not move to the rally point when taking off from the repair bay.

`Aircraft.GetActorBelow()` only looks for Reservable actors, and Reservable in its current form won't mesh with repair-only hosts. I'd like to leave this as "known issue, but not serious enough to be a release blocker" and adress it properly during the next release cycle.

> I also noticed that when an aircraft with TakeOffOnResupply == true is made to resupply using the force-enter command and the stop command is given it will forget to supress the TakeOff. This is due to the Resupply activity being reqeued without the stayOnResupplier parameter.

Fixing this properly would likely require to remove the hacky re-queueing of `Resupply`, which in turn would require `Attack*` traits to only cancel their own activities but not the entire queue on "Stop" orders, which would require us to refactor `Actor.CancelActivity()`.
tl;dr: Another issue I'd like to defer to Next+1.

tovl commented:
I agree with deferring these issues to Next+1.

As far as refactoring `Actor.CancelActivity()` is concerned, I already have a potential solution in mind that would not require us touching stop orders or requeueing `Resupply` but would require reworking `Actor.QueueActivity()`. I will try that out as soon as this is merged.

tovl commented:
Found another bug: In D2K, orders queued after the repair order get ignored when the unit has been carried by a carryall.

reaperrr commented:
I'm pretty sure it was that way before this PR already.
I could pass a sequence of `Resupply, NextActivity` instead of only `Resupply` through `RequestTransport`, not sure if that also implicitly passes any activities queued after `NextActivity`, though.

tovl commented:
Any activity in the queue carries a reference to the next activity in the queue. So passing `NextActivity` would also implicitly pass `NextActivity.NextActivity`, `NextActivity.NextActivity.Nextactivity`, etc..., so I suspect this will work.

reaperrr commented:
OK, not that easy, unfortunately. Resupply triggers the request right on its first tick, so any activity queued after the transport was requested is ignored. I think this might require refactoring how RequestTransport works, but that would be out of scope for this PR.

pchote commented:
>  I think this might require refactoring how RequestTransport works, but that would be out of scope for this PR.

When the time comes, this can be done the same way I was arguing the transport locking for passengers should be done. Right now the carryall makes the requesting actor cancel its activity and queues a `WaitFor` until it has picked it up. Instead, `Carryable` should grant a condition that pauses `Mobile` so that the activity queue doesn't need to be touched at all.

reaperrr commented:
Can we at least leave that to a follow-up? I'm literally getting a headache right now so I won't get that done today, and as long as this PR drags on I won't make progress on #16695's update rule, either.

pchote commented:
"When the time comes" are the magic words :smile: IMO that is a job for Next + 1.

reaperrr commented:
I was referring to the "right now" as well ;)

tovl reviewed (dismissed):
With those known issues deferred to the future, I think this is good to go. :+1:

teinarss commented on OpenRA.Mods.Common/Activities/Resupply.cs:
```diff
@@ -33,6 +33,7 @@ public class Resupply : Activity
 		readonly IMove move;
 		readonly Aircraft aircraft;
 		readonly bool stayOnResupplier;
+		readonly bool wasRepair;
```
nitpick: wasRepaired

teinarss reviewed (commented):
Otherwise it looks good. Did not test the update rules

reaperrr commented:
Fixed.

pchote reviewed (approved):
The important issues from earlier reviews appear to have been resolved.

The logic (both code and expected gameplay behaviours) here is complicated, so I expect that more regressions will probably be found in this code. The changes here are a needed step forward, so lets get these in as a first step to all the other fixes discussed above, and any other fixes that might be needed for Next.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/bf379b06d37d013355fb28563aaa4ce4263b801b
